// lede/dist/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const glob = require("glob");
const fs_extra_1 = require("fs-extra");
function copyProm(src, targ) {
    return new Promise((resolve, reject) => {
        fs_extra_1.copy(src, targ, { clobber: true }, (err) => {
            if (err) {
                return reject(err);
            }
            return resolve();
        });
    });
}
exports.copyProm = copyProm;
function globProm(path, cwd) {
    return new Promise((resolve, reject) => {
        glob(path, {
            cwd: cwd ? cwd : process.cwd()
        }, (err, paths) => {
            if (err) {
                return reject(err);
            }
            return resolve(paths);
        });
    });
}
exports.globProm = globProm;
function existsProm(path) {
    return new Promise((resolve, reject) => {
        fs_extra_1.stat(path, (err, stats) => {
            if (err)
                return reject(err);
            return resolve({ file: stats.isFile(), dir: stats.isDirectory() });
        });
    });
}
exports.existsProm = existsProm;
function createDir(path) {
    return new Promise((resolve, reject) => {
        fs_extra_1.ensureDir(path, (err) => {
            if (err) {
                return reject(err);
            }
            return resolve();
        });
    });
}
exports.createDir = createDir;
function asyncMap(array, f) {
    return __awaiter(this, void 0, Promise, function* () {
        let returns = [];
        for (let item of array) {
            returns.push(yield f(item));
        }
        return returns;
    });
}
exports.asyncMap = asyncMap;
function writeProm(data, file) {
    return new Promise((resolve, reject) => {
        fs_extra_1.writeFile(file, data, (err) => {
            if (err) {
                return reject(err);
            }
            return resolve();
        });
    });
}
exports.writeProm = writeProm;
function readJsonProm(path) {
    return new Promise((resolve, reject) => {
        fs_extra_1.readJson(path, (err, r) => {
            if (err) {
                return reject(err);
            }
            return resolve(r);
        });
    });
}
exports.readJsonProm = readJsonProm;
