// lede/dist/lede/Lede.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const DependencyAssembler_1 = require("./DependencyAssembler");
const CacheBuilder_1 = require("./CacheBuilder");
class Lede {
    constructor(workingDir, compilers, deployers, logger) {
        this.workingDir = workingDir;
        this.compilers = compilers;
        this.deployers = deployers;
        this.logger = logger;
    }
    deploy(deployer, debug = true, pr) {
        return __awaiter(this, void 0, Promise, function* () {
            if (!pr) {
                pr = yield Lede.assembleDeps(this.workingDir, this.logger);
            }
            pr.context.$debug = debug;
            yield Lede.buildCache(pr, this.logger);
            let compiledPage = yield Lede.compilePage(this.compilers, pr, this.logger);
            deployer = this.deployers[deployer];
            yield Lede.deployPage(deployer, pr, compiledPage, this.logger);
            return pr;
        });
    }
    static deployPage(deployer, projReport, compiledPage, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            logger.debug({ deployer: deployer, projReport: projReport, compiledPage: compiledPage });
            logger.info("Deploying project.");
            try {
                yield deployer.deploy({ report: projReport, compiledPage: compiledPage });
            }
            catch (e) {
                logger.error({ err: e }, "Error deploying project.");
            }
        });
    }
    static compilePage(compilers, proj, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            logger.info("Compiling project.");
            logger.debug({ projectReport: proj });
            try {
                return yield compilers.html.compile(proj, { css: compilers.css, js: compilers.js });
            }
            catch (e) {
                logger.error({ err: e }, "Error compiling page.");
            }
        });
    }
    static assembleDeps(workingDir, logger) {
        return __awaiter(this, void 0, Promise, function* () {
            try {
                logger.debug({ workingDir: workingDir });
                logger.info("Assembling dependencies");
                let da = new DependencyAssembler_1.DependencyAssembler(workingDir);
                return yield da.assemble();
            }
            catch (e) {
                logger.error({ err: e });
                if (e.code === "CircularDepError") {
                    logger.info(`${e.message} depends on a project which depends on itself.`);
                }
                else if (e.code === "NotAFile") {
                    logger.info(`${e.message} is not a file.`);
                }
            }
        });
    }
    static buildCache(proj, logger) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                logger.info("Caching assets");
                let cb = new CacheBuilder_1.CacheBuilder(proj);
                yield cb.buildCache();
            }
            catch (e) {
                logger.error({ err: e }, "Error creating cache");
            }
        });
    }
    static buildProject(logger, projectReport) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof projectReport === "string") {
                projectReport = yield Lede.assembleDeps(projectReport, logger);
            }
            yield Lede.buildCache(projectReport, logger);
        });
    }
}
exports.Lede = Lede;
