// lede/dist/compilers/Es6Compiler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const browserify = require("browserify");
const babelify = require("babelify");
const path_1 = require("path");
const glob = require("glob");
function globProm(path, cwd) {
    return new Promise((resolve, reject) => {
        glob(path, {
            cwd: cwd ? cwd : process.cwd()
        }, (err, paths) => {
            if (err) {
                return reject(err);
            }
            return resolve(paths);
        });
    });
}
class Es6Compiler {
    constructor(options) {
        this.options = options;
    }
    ;
    compile(report, usedBits) {
        return __awaiter(this, void 0, void 0, function* () {
            let requireable = yield Es6Compiler.getRequirablePaths(path_1.resolve(report.workingDirectory, ".ledeCache", "scripts"));
            let globals = yield Es6Compiler.bundleGlobals(report, requireable);
            let bits = yield Es6Compiler.bundleBits(report, requireable, usedBits);
            return {
                bits: bits,
                globals: globals
            };
        });
    }
    static bundleBits(report, requireable, bits) {
        return __awaiter(this, void 0, void 0, function* () {
            let bitPaths = bits.map(b => path_1.resolve(report.workingDirectory, ".ledeCache", "bits", b, "interact.js"));
            return Es6Compiler.bundleProm(bitPaths, requireable, path_1.resolve(report.workingDirectory, ".ledeCache", "scripts"), report.context.$debug);
        });
    }
    static bundleGlobals(report, requireable) {
        return __awaiter(this, void 0, void 0, function* () {
            let globalPaths = report.scripts.map(s => path_1.resolve(report.workingDirectory, ".ledeCache", "scripts", s));
            return yield Es6Compiler.bundleProm(globalPaths, requireable, path_1.resolve(report.workingDirectory, ".ledeCache", "scripts"), report.context.$debug);
        });
    }
    static bundleProm(toAdd, requireable, cwd, debug) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                let b = browserify({ basedir: cwd, debug: debug });
                b.require(requireable);
                b.add(toAdd);
                b.transform(babelify, { presets: require('babel-preset-es2015') });
                b.bundle((err, res) => {
                    if (err)
                        return reject(err);
                    return resolve(res.toString());
                });
            });
        });
    }
    static getRequirablePaths(searchDir) {
        return __awaiter(this, void 0, void 0, function* () {
            let paths = [];
            let projects = yield globProm(`*`, searchDir);
            for (let proj of projects) {
                let projName = path_1.basename(proj);
                let files = yield globProm(`*`, path_1.resolve(searchDir, proj));
                for (let file of files) {
                    let fileName = path_1.basename(file);
                    paths.push({
                        file: path_1.resolve(searchDir, projName, fileName),
                        expose: `${projName}/${fileName}`
                    });
                }
            }
            return paths;
        });
    }
}
exports.Es6Compiler = Es6Compiler;
