// lede/dist/compilers/SassCompiler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const fs_extra_1 = require("fs-extra");
const node_sass_1 = require("node-sass");
const path_1 = require('path');
function asyncMap(array, f) {
    return __awaiter(this, void 0, Promise, function* () {
        let returns = [];
        for (let item of array) {
            returns.push(yield f(item));
        }
        return returns;
    });
}
class SassCompiler {
    constructor(options = { includePaths: [], outputStyle: 'compact', sourceComments: false, sourceMapEmbed: false }) {
        this.options = options;
    }
    compile(report, bits) {
        return __awaiter(this, void 0, void 0, function* () {
            this.options.includePaths.push(path_1.join(report.workingDirectory, '.ledeCache', 'styles'));
            let compiledGlobals = yield SassCompiler.compileGlobals(report, Object.assign({}, this.options));
            let compiledBits = yield SassCompiler.compileBits(report, Object.assign({}, this.options), bits);
            return {
                globals: compiledGlobals,
                bits: compiledBits
            };
        });
    }
    static compileBits(report, options, bits) {
        return __awaiter(this, void 0, void 0, function* () {
            let bitPaths = bits.map(b => path_1.join(report.workingDirectory, '.ledeCache', 'bits', b, 'style.scss'));
            let compiledBits = yield asyncMap(bitPaths, (b) => __awaiter(this, void 0, void 0, function* () {
                return yield SassCompiler.renderFile(options, b);
            }));
            return compiledBits.join('');
        });
    }
    static compileGlobals(report, options) {
        return __awaiter(this, void 0, void 0, function* () {
            let stylesDir = path_1.join(report.workingDirectory, '.ledeCache', 'styles');
            let opts = Object.assign({}, options);
            let styleSheets = yield asyncMap(report.styles, (f) => __awaiter(this, void 0, void 0, function* () {
                return yield SassCompiler.renderFile(opts, path_1.join(stylesDir, f));
            }));
            return styleSheets.join('');
        });
    }
    static renderFile(options, filePath) {
        let stream = fs_extra_1.createReadStream(filePath);
        let data = "";
        return new Promise((resolve, reject) => {
            stream.on('data', d => data += d.toString());
            stream.on('end', () => {
                if (!data) {
                    return resolve("");
                }
                node_sass_1.render({
                    data: data,
                    includePaths: options.includePaths,
                    outputStyle: options.outputStyle,
                    sourceComments: options.sourceComments,
                    sourceMapEmbed: options.sourceMapEmbed
                }, (err, res) => {
                    if (err) {
                        return reject(err);
                    }
                    return resolve(res.css.toString());
                });
            });
        });
    }
}
exports.SassCompiler = SassCompiler;
