// lede/dist/deployers/S3Deployer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const aws_sdk_1 = require("aws-sdk");
const path_1 = require("path");
const FileSystemDeployer_1 = require("./FileSystemDeployer");
const utils_1 = require("../utils");
const fs_extra_1 = require("fs-extra");
class S3Deployer extends FileSystemDeployer_1.FileSystemDeployer {
    constructor(deployDir, bucket, projName) {
        super(deployDir);
        this.bucket = bucket;
        this.projName = projName;
    }
    deploy(project) {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            yield _super("deploy").call(this, project);
            let toUpload = yield utils_1.globProm('**/*.*', this.deployDir);
            for (let file of toUpload) {
                yield S3Deployer.uploadToS3({ Bucket: this.bucket, filePath: path_1.resolve(this.deployDir, file), Key: `${this.projName}/${file}` });
            }
        });
    }
    static uploadToS3({ Bucket, Key, filePath }) {
        let contentType = null;
        let fileType = path_1.basename(filePath).split('.')[path_1.basename(filePath).split('.').length - 1];
        switch (fileType) {
            case 'json':
                contentType = "application/json";
                break;
            case 'css':
                contentType = "text/css";
                break;
            case 'jpg':
                contentType = "image/jpeg";
                break;
            case 'png':
                contentType = "image/png";
                break;
            case 'html':
                contentType = "text/html";
                break;
            case 'js':
                contentType = "application/javascript";
                break;
            default:
                contentType = "text/plain";
                break;
        }
        return new Promise((resolve, reject) => {
            let upload = new aws_sdk_1.S3.ManagedUpload({
                params: {
                    Bucket: Bucket,
                    Key: Key,
                    Body: fs_extra_1.createReadStream(filePath),
                    ACL: "bucket-owner-full-control",
                    ContentType: contentType
                }
            });
            upload.send((err, data) => {
                if (err) {
                    return reject(err);
                }
                return resolve();
            });
        });
    }
}
exports.S3Deployer = S3Deployer;
